### PR TITLE
[For discussion] Factor out election.Runner

### PR DESCRIPTION
**Pull request by daviddrysdale**
This just attempts to pull out the `electionRunner` code from `LogOperationManager`, as far as possible without making any other changes to it.

Changes that were necessary:
 - add a notification callback to `MasterTracker` for triggering metric updates
 - move resignation metric tracking to point of resignation
 - add an exposed `ID` field in the `resignation` struct.

### CODE REVIEW

File: util/election/runner.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	MasterHoldInterval time.Duration
+	// ResignOdds gives the chance of resigning mastership after each
+	// check interval, as the N for 1-in-N.
+	ResignOdds int
+
+	TimeSource util.TimeSource
+}
+
+// fixupRunnerConfig ensures operation parameters have required minimum values.
+func fixupRunnerConfig(cfg *RunnerConfig) {
```
Comment (pavelkalinnikov) [inline util/election/runner.go]: I think that these values should be application- and/or (MasterElection)implementation- specific.
It's fine having them here though, because of this PR's purpose.

File: util/election/runner.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	return &Runner{
+		Cancel:   cancel,
+		id:       id,
+		cfg:      cfg,
+		tracker:  tracker,
+		election: el,
+	}
+}
+
+// Run performs the process of performing a continuous election. It runs continuously until
```
Comment (pavelkalinnikov) [inline util/election/runner.go]: Could you rephrase it to avoid "performs performing"? E.g.:
```
// Run continuously takes part in master election on behalf of this instance.
// Returns when the context is canceled or an internal error is encountered.
```

File: util/election/runner.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	}
+	return rand.Intn(er.cfg.ResignOdds) == 0
+}
+
+// Resignation indicates that a master should explicitly resign mastership, by invoking
+// the Execute() method at a point where no master-related activity is ongoing.
+type Resignation struct {
+	ID   int64
+	er   *Runner
+	done chan<- bool
```
Comment (pavelkalinnikov) [inline util/election/runner.go]: Could make this `chan struct{}` and just close it in `Execute`.

File: util/election/runner.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply util/election/runner.go]: Yep, I think so.

File: util/election/runner.go (hunk 2 above)
Comment (daviddrysdale) [reply util/election/runner.go]: Done.

File: util/election/runner.go (hunk 3 above)
Comment (daviddrysdale) [reply util/election/runner.go]: We could indeed, but that would contradict the spirit of this PR.

**Comment by daviddrysdale** [comment]
So I guess my question is: does this help with clarifying the behaviour changes in #1177 ?   If that PR were rebased on top of this (which is admittedly a pain), could we then see the API and internal changes to the election running process?

File: util/election/runner.go (hunk 3 above)
Comment (pavelkalinnikov) [reply util/election/runner.go]: SGTM, let's leave it as is.

**Comment by pavelkalinnikov** [comment]
This is a really good question. The effect of this PR is positive anyway, I think, so we should probably proceed here.

With regards to #1177 I am not so sure if a pure rebase would help to nicely isolate the changes. That one probably can be further split:
- Replace `ResignOdds/ShouldResign` code by a less restrictive `ResignSpread` option.
- Move resignation decision from `Runner` to higher level.
- Make `MasterTracker` optional in `Runner`, or fully decouple it.
- Finally convert the `Runner` to something similar to #1177 if needed (atm it's hard to tell), plus tests.
